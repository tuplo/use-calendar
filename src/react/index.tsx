import type { IUseCalendarOptions } from "../index";
import { useCalendar } from "../index";

function Calendar(props?: Partial<IUseCalendarOptions>) {
	const { months, getDayProps, getPrevMonthProps, getNextMonthProps } =
		useCalendar(props);

	return (
		<>
			{months.map(({ year, month, weeks }) => (
				<div key={`${year}${month}`}>
					<header>
						<h1>
							{month} {year}
						</h1>
					</header>
					<nav>
						<button {...getPrevMonthProps()}>Prev</button>
						<button {...getNextMonthProps()}>Next</button>
					</nav>
					{weeks.map((week, weekIndex) =>
						week.map((day, dayIndex) =>
							day ? (
								<button
									key={`${year}${weekIndex}${dayIndex}`}
									{...getDayProps({ day })}
								>
									{day.date.getDate()}
								</button>
							) : (
								<span key={`${year}${weekIndex}${dayIndex}`} />
							)
						)
					)}
				</div>
			))}
		</>
	);
}

export default Calendar;
