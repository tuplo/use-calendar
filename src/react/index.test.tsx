import "@testing-library/jest-dom";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";

import Calendar from "./index";

describe("useCalendar UI", () => {
	const user = userEvent.setup();

	it("renders", () => {
		const { container } = render(<Calendar />);

		expect(container).toMatchSnapshot();
		expect(screen.getByText("6 2022")).toBeInTheDocument();
	});

	describe("onDateSelected", () => {
		const dateNowSpy = vi
			.spyOn(Date, "now")
			.mockReturnValue(new Date("2022-07-25").getTime());

		afterAll(() => {
			dateNowSpy.mockRestore();
		});

		it("calls provided handler when user clicks day", async () => {
			const onDateSelectedSpy = vi.fn();
			render(<Calendar onDateSelected={onDateSelectedSpy} />);
			await act(async () => {
				await user.click(screen.getByText("12"));
			});

			const expected = {
				date: new Date("2022-07-12T00:00:00.000Z"),
				isSelectable: true,
			};
			expect(onDateSelectedSpy).toHaveBeenCalledTimes(1);
			expect(onDateSelectedSpy).toHaveBeenCalledWith(expected);
		});

		it("doesn't call when user clicks not available dates", async () => {
			const onDateSelectedSpy = vi.fn();
			const availableDates = [10, 12, 13, 14, 15].map(
				(date) => new Date(`2022-07-${date}`)
			);
			render(
				<Calendar
					onDateSelected={onDateSelectedSpy}
					availableDates={availableDates}
				/>
			);
			await user.click(screen.getByText("25"));

			expect(onDateSelectedSpy).toHaveBeenCalledTimes(0);
		});
	});
});
