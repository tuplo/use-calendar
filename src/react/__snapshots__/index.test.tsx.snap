// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useCalendar UI > renders 1`] = `
<div>
  <div>
    <header>
      <h1>
        6
         
        2022
      </h1>
    </header>
    <nav>
      <button
        aria-label="Go back 1 month"
        role="button"
        type="button"
      >
        Prev
      </button>
      <button
        aria-label="Go forward 1 month"
        role="button"
        type="button"
      >
        Next
      </button>
    </nav>
    <span />
    <span />
    <span />
    <span />
    <span />
    <button
      aria-label="July 1, 2022"
      role="button"
    >
      1
    </button>
    <button
      aria-label="July 2, 2022"
      role="button"
    >
      2
    </button>
    <button
      aria-label="July 3, 2022"
      role="button"
    >
      3
    </button>
    <button
      aria-label="July 4, 2022"
      role="button"
    >
      4
    </button>
    <button
      aria-label="July 5, 2022"
      role="button"
    >
      5
    </button>
    <button
      aria-label="July 6, 2022"
      role="button"
    >
      6
    </button>
    <button
      aria-label="July 7, 2022"
      role="button"
    >
      7
    </button>
    <button
      aria-label="July 8, 2022"
      role="button"
    >
      8
    </button>
    <button
      aria-label="July 9, 2022"
      role="button"
    >
      9
    </button>
    <button
      aria-label="July 10, 2022"
      role="button"
    >
      10
    </button>
    <button
      aria-label="July 11, 2022"
      role="button"
    >
      11
    </button>
    <button
      aria-label="July 12, 2022"
      role="button"
    >
      12
    </button>
    <button
      aria-label="July 13, 2022"
      role="button"
    >
      13
    </button>
    <button
      aria-label="July 14, 2022"
      role="button"
    >
      14
    </button>
    <button
      aria-label="July 15, 2022"
      role="button"
    >
      15
    </button>
    <button
      aria-label="July 16, 2022"
      role="button"
    >
      16
    </button>
    <button
      aria-label="July 17, 2022"
      role="button"
    >
      17
    </button>
    <button
      aria-label="July 18, 2022"
      role="button"
    >
      18
    </button>
    <button
      aria-label="July 19, 2022"
      role="button"
    >
      19
    </button>
    <button
      aria-label="July 20, 2022"
      role="button"
    >
      20
    </button>
    <button
      aria-label="July 21, 2022"
      role="button"
    >
      21
    </button>
    <button
      aria-label="July 22, 2022"
      role="button"
    >
      22
    </button>
    <button
      aria-label="July 23, 2022"
      role="button"
    >
      23
    </button>
    <button
      aria-label="July 24, 2022"
      role="button"
    >
      24
    </button>
    <button
      aria-label="July 25, 2022"
      role="button"
    >
      25
    </button>
    <button
      aria-label="July 26, 2022"
      role="button"
    >
      26
    </button>
    <button
      aria-label="July 27, 2022"
      role="button"
    >
      27
    </button>
    <button
      aria-label="July 28, 2022"
      role="button"
    >
      28
    </button>
    <button
      aria-label="July 29, 2022"
      role="button"
    >
      29
    </button>
    <button
      aria-label="July 30, 2022"
      role="button"
    >
      30
    </button>
    <button
      aria-label="July 31, 2022"
      role="button"
    >
      31
    </button>
    <span />
    <span />
    <span />
    <span />
    <span />
    <span />
  </div>
</div>
`;
